File: java_version/Czlowiek.java
public class Czlowiek extends Zwierze {
    private int cooldown;
    private int czasTrwania;
    private boolean aktywnaUmiejetnosc;
    private int silaBazowa;
    private int dx = 0;
    private int dy = 0;

    public Czlowiek(int x, int y, Swiat swiat) {
        super(5, 4, x, y, swiat);
        this.silaBazowa = this.sila;
    }

    @Override
    public char rysowanie() {
        return 'H';
    }

    @Override
    public String getNazwa() {
        return "Czlowiek";
    }

    @Override
    public Organizm stworzPotomka(int nowyX, int nowyY) {
        return null; // Człowiek się nie rozmnaża
    }

    @Override
    public void akcja() {
        if (aktywnaUmiejetnosc) {
            if (czasTrwania == 5) {
                sila = 10;
                swiat.dodajLog("Eliksir aktywowany! Sila Czlowieka wynosi 10.");
            } else {
                sila = silaBazowa + (5 - czasTrwania);
            }

            czasTrwania--;

            if (czasTrwania == 0) {
                aktywnaUmiejetnosc = false;
                sila = silaBazowa;
                cooldown = 5;
                swiat.dodajLog("Eliksir wygasl. Sila Czlowieka wrocila do " + silaBazowa);
            }
        } else if (cooldown > 0) {
            cooldown--;
        }

        if (dx != 0 || dy != 0) {
            int nowyX = x + dx;
            int nowyY = y + dy;
            if (swiat.czyWPlanszy(nowyX, nowyY)) {
                wykonajRuch(nowyX, nowyY);
            }
        }

        dx = 0;
        dy = 0;
    }

    @Override
    public void kolizja(Organizm inny) {
        super.kolizja(inny);
    }

    @Override
    public boolean czyTenSamGatunek(Organizm inny) {
        return false;
    }

    public void aktywujUmiejetnosc() {
        if (!aktywnaUmiejetnosc && cooldown == 0) {
            aktywnaUmiejetnosc = true;
            czasTrwania = 5;
            swiat.dodajLog("Czlowiek wypil magiczny eliksir");
        } else {
            swiat.dodajLog("Eliksir niedostepny (cooldown: " + cooldown + ")");
        }
    }

    public void obsluzSterowanie(int key) {
        switch (key) {
            case 38: dx = 0; dy = -1; break; // Góra (VK_UP)
            case 40: dx = 0; dy = 1; break;  // Dół
            case 37: dx = -1; dy = 0; break; // Lewo
            case 39: dx = 1; dy = 0; break;  // Prawo
            case 'U': case 'u': aktywujUmiejetnosc(); break;
            default: break;
        }
    }

    public int getCooldown() { return cooldown; }
    public boolean isAktywnaUmiejetnosc() { return aktywnaUmiejetnosc; }
    public int getCzasTrwania() { return czasTrwania; }

    public void setCooldown(int c) { this.cooldown = c; }
    public void setAktywnaUmiejetnosc(boolean a) { this.aktywnaUmiejetnosc = a; }
    public void setCzasTrwania(int t) { this.czasTrwania = t; }
}

File: cpp_version/Czlowiek.cpp
#include "Czlowiek.h"
#include "Swiat.h"
#include <conio.h>

Czlowiek::Czlowiek(int x, int y, Swiat& swiat)
	: Zwierze(5, 4, x, y, swiat),
	cooldown(0), czasTrwania(0), aktywnaUmiejetosc(false), silaBazowa(sila), dx(0), dy(0){}

char Czlowiek::rysowanie() const {
	return 'H';
}

std::string Czlowiek::getNazwa() const {
	return "Czlowiek";
}

std::shared_ptr<Organizm> Czlowiek::stworzPotomka(int, int) {
	return nullptr;
}

void Czlowiek::akcja() {
	if (aktywnaUmiejetosc) {
		if (czasTrwania == 5) {
			sila = 10;
			swiat.dodajLog("Eliksir aktywowany! Sila Czlowieka wynosi 10.");
		}
		else {
			sila = silaBazowa + (5 - czasTrwania);
		}
		czasTrwania--;
		if (czasTrwania == 0) {
			aktywnaUmiejetosc = false;
			sila = silaBazowa;
			cooldown = 5;
			swiat.dodajLog("Eliksir wygasl.Sila czlowieka wrocila do " + std::to_string(silaBazowa));
		}
	}
	else if (cooldown > 0) {
		cooldown--;
	}
	if (dx != 0 || dy != 0) {
		int nowyX = x + dx;
		int nowyY = y + dy;
		if (nowyX >= 0 && nowyX < swiat.getSzerokosc() && nowyY >= 0 && nowyY < swiat.getWysokosc()) {
			wykonajRuch(nowyX, nowyY);
		}
	}
	dx = 0;
	dy = 0;
}

void Czlowiek::kolizja(Organizm& inny) {
	Zwierze::kolizja(inny);
}

void Czlowiek::aktywujUmiejetnosc() {
	if (!aktywnaUmiejetosc && cooldown == 0) {
		aktywnaUmiejetosc = true;
		czasTrwania = 5;
		swiat.dodajLog("Czlowiek wypil magiczny eliksir");
	}
	else {
		swiat.dodajLog("Eliksir niedostepny (cooldown: " + std::to_string(cooldown)+ ")");
	}
}


void Czlowiek::obsluzSterowanie(Czlowiek& czlowiek, int key) {
	switch (key)
	{
	case 72: dx = 0; dy = -1; break; //w gore
	case 80: dx = 0; dy = 1; break;  //w dol
	case 75: dx = -1; dy = 0; break; //w lewo
	case 77: dx = 1; dy = 0; break;  //w prawo
	case 'u': aktywujUmiejetnosc(); break;
	default: break;
	}
}

bool Czlowiek::czyTenSamGatunek(Organizm&) const {
	return false;
}

int Czlowiek::getCooldown() const {
	return cooldown;
}

bool Czlowiek::isAktywnaUmiejetosc() const {
	return aktywnaUmiejetosc;
}

int Czlowiek::getCzasTrwania() const {
	return czasTrwania;
}

void Czlowiek::setCooldown(int c) {
	cooldown = c;
}

void Czlowiek::setAktywnaUmiejetnosc(bool a) {
	aktywnaUmiejetosc = a;
}

void Czlowiek::setCzasTrwania(int t) {
	czasTrwania = t;
}